doc/api/packages.md

```diff
@@ -83,6 +85,41 @@ future-proof the package in case the default type of Node.js ever changes, and
 it will also make things easier for build tools and loaders to determine how the
 files in the package should be interpreted.
 
+### Modules loaders
+
+Node.js has two system for resolving a specifier and load modules.
+
+There is the CommonJS module loader:
+
+* It is fully synchronous.
+* It is monkey patchable.
+* When resolving a specifier, if no exact match is found, it will try to add
+  extensions (`.js`, `.json`, and finally `.node`).
+* It supports [folders as modules][].
+* It treats `.json` as JSON text files.
+* `.node` files are interpreted as compiled addon modules loaded with
+  `process.dlopen()`.
+* It treats files that do not have `.json` or `.node` extension as JavaScript
+  text files.
+* It cannot be used to load ECMAScript modules. Attempting to do so will result
+  in a [`ERR_REQUIRE_ESM`][] error.
+* It can be accessed using `require` function.
```

Comment:
```suggestion
* It allows one CommonJS module to load another using the `require` function.
* It allows CommonJS modules to asynchronously import ECMAScript modules,
  using `import()` expressions.
* `require` cannot be used to load ECMAScript modules. Attempting to do so
  will result in a [`ERR_REQUIRE_ESM`][] error.
```

Reply:
I don't know if you noticed, but I tried to list the same features in both lists, so user can compare the difference between CJS and ESM loaders. If we add an item here, can we add it to the ESM list as well?

Reply:
You could combine the two `require`-related ones, maybe?

```md
* It allows one CommonJS module to load another using the `require` function.
  `require` cannot be used to load ECMAScript modules. Attempting to do so
  will result in a [`ERR_REQUIRE_ESM`][] error.
* It allows CommonJS modules to asynchronously import ECMAScript modules,
  using `import()` expressions.
```

Reply:
> ```lisp
> It allows one CommonJS module to load another using the `require` function.
> ```

It also allows ES module to load a CJS module using `require` function. I'm not sure this information is useful tbh. Is it correct that the information you are trying to convey is that all JS files are treated as CJS? If so sure I can add that.

> ```lisp
> It allows CommonJS modules to asynchronously import ECMAScript modules,
>   using `import()` expressions.
> ```

I don't see how this is related to the CJS loader. `import()` is exposed in CJS modules, it is explicitly called out is several places, but I think it would add too much confusion to list `import()` as a CJS loader feature.

Reply:
I went with https://github.com/nodejs/node/blob/21b45bf922d2768ce10d480a9c843ec20b92d9f3/doc/api/packages.md#L103-L107 wdyt?

Reply:
The “it” here I think refers to the CommonJS _loader,_ but I think what users care about is what can be done from a CommonJS _context._ As in, you _can_ load an ES module from within a CommonJS module; you just need to use `import()`. And you can load CommonJS modules from within ES modules, either via `import` or by creating `require`. In those cross-module-system cases we sort of have one loader using the other. So while it may technically be correct that the CommonJS loader only handles CommonJS and the ESM loader only handles ESM, that implies that the two systems can’t interoperate which is _not_ an impression we want to convey.

What I was trying to get across was that `require` can only refer to CommonJS, `import` and `import()` can refer to either module system; and CommonJS supports `import()`. As in, _these are the ways that interoperability is possible._ It’s important to note that `require` doesn’t support ESM, and it’s just as important to suggest `import()` is the solution for that.

Reply:
Fair, I agree we don't want to add this kind of possible confusion. What do you think of this: https://github.com/nodejs/node/blob/93d6a233cf3afb504b62c6de9396e9edccd1da57/doc/api/packages.md#L105-L108

Reply:
I guess stepping back for a second, what’s the value of stating that the CommonJS loader only loads CommonJS? That’s describing Node’s internal workings which may be technically correct, but easy to misunderstand; and maybe it’s better just left out entirely.

Maybe just include the latter part, in which case you can get closer to your original version:

```markdown
 * It tries to load all JavaScript text files as CommonJS modules. If such a file
   is an ECMAScript module, an [`ERR_REQUIRE_ESM`][] error is thrown.
   Within a CommonJS module, [`import()`][] can be used to invoke the
   ES module loader to import an ES module.
```

Reply:
What I don't "like" with this last version is it gives the impression that if an ESM file do not use any ESM specific syntax, CJS loader would (try to) parse it as CJS. Can we have a version without using `tries to`?

Do we really need to document here that `ERR_REQUIRE_ESM` is thrown? It's already documented in `modules.md`, I'd prefer if we leave it out to simplify that list item.

> what’s the value of stating that the CommonJS loader only loads CommonJS?

ESM loader can load more than just ESM, so it may not be obvious that CJS only loads CJS?

Reply:
> ESM loader can load more than just ESM, so it may not be obvious that CJS only loads CJS?

Can it though? Doesn’t it just call the CommonJS loader to handle cases of `import` of CJS? Just like how the CommonJS loader calls the ESM loader to handle cases of `import()` of ESM?

We can leave out the error message, sure. But then I don’t think there’s anything left of this bullet point, and so we should just remove it. Both loaders use the other to handle the other system’s modules. The only thing worth mentioning is how `import`/`import()` can handle either module system while `require` only supports ESM. That’s a really important point that we should make if we’re not doing so already, and is much more valuable to users than a description of how the internals of the loaders work.

Reply:
(`require` only supports CJS)

Reply:
IMHO there's a clear difference: with `Module._load` there is no way to load a file as ESM; with `esmLoader.load`, you can load a file as CJS. That's what I'm talking about, maybe it's not an information relevant for end users, but I believe it is for loader hooks authors.
`require` and `import` are user facing features, and are already documented elsewhere, it's easier if we define `require => CJS loader` and `import => ESM loader` (and it's mostly true), and I think it would be detrimental to imply that `import` and `require` behave differently depending on the parse goal of the file.

I'm going to land as is, hopefully we can improve this part of the docs in future PRs.

Reply:
> ESM loader can load more than just ESM, so it may not be obvious that CJS only loads CJS?

But can it really? If you `import` a CommonJS file, isn’t it just passing that off to the CommonJS loader for loading? So in that sense, each loader only handles its domain.

CommonJS _modules_ load more than just CommonJS, and the inverse for ES modules, and that’s what matters to end users. Stating that the _loader_ can only handle CommonJS just adds confusion.

I think that matters to users is that `require` can only take CommonJS, while `import`/`import()` can take either. And that `import` doesn’t exist in CommonJS, but `import()` does.

All this other stuff about what the loaders can do is unnecessary detail that can confuse people. I understand the desire to keep the whole list focused on the CommonJS loader, which is why I’m thinking it’s best to just leave out this detail unless you want to make this bullet about `require` vs `import` rather than about the loader itself.